docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+* Ability to add non-error notifications from the build process
+* Add extra metadata associated to the notification: icon, color, header, body, etc
+* Support different types of notifications (e.g. error, warning, note, tip)
+* Re-use the new notification system for product updates (e.g. new features, deprecated config keys)
+* Message content lives on Python classes that can be translated and formatted with objects (e.g. Build, Project)
+* Message could have richer content (e.g. HTML code) to generate links and emphasis
+* Notifications have trackable state (e.g. unread (default)=never shown, read=shown, dismissed=don't show again)
+* An object (e.g. Build, Organization) can have more than 1 notification attached
+* Remove hardcoded notifications from the templates
+* Notifications can be attached to Project, Organization, Build and User models
+* Specific notifications can be shown under the user's bell icon
+* Easy way to cleanup notification on status changes (e.g. subscription failure notification is auto-deleted after CC updated)
+* Notifications attached to Organization/Project dissappear for all the users once they are dismissed by anyone
+
+
+Non-goals
+---------
+
+* Create new Build "state" or "status" option for these fields
+* Implement the new notification in the old dashboard
+* Define front-end code implementation
+* Replace email or webhook notifications
+
+
+Small notes and other considerations
+------------------------------------
+
+* Django message system is not enough for this purpose.
+* Use a new model to store all the required data (expandable in the future)
+* How do we handle translations?
+  We should use ``_("This is the message shown to the user")`` in Python code and return the proper translation when they are read.
+* Reduce complexity on ``Build`` object (remove ``Build.status`` and ``Build.error`` fields among others).
+* Since the ``Build`` object could have more than 1 notification, when showing them, we will sort them by importane: errors, warnings, note, tip.
+* In case we need a pretty specific order, we can add an extra field for that, but it adds unnecessary complexity at this point.
+* For those notifications that are attached to the ``Project`` or ``Organization``, should it be shown to all the members even if they don't have admin permissions?
+  If yes, this is good because all of them will be notified but only some of them will be able to take an action.
+  If no, non-admin users won't see the notification and won't be able to communicate this to the admins
+* Notification could be attached to a ``BuildCommand`` in case we want to display a specific message on a command itself.
+  We don't know how useful this will be, but it's something we can consider in the future.
+* Notification preferences: what kind of notifications I want to see in my own bell icon?
+
+  - Build errors
+  - Build tips
+  - Product updates
+  - Blog post news
+  - Organization updates
+  - Project updates
```

Comment:
This is helpful, I wasn't think of using this area for some of these notification types, but I see how some of these would fit really well.

The only one I'd discard from the list right away is build errors. These would overwhelm this area, especially any project that has frequent build errors. A user that logs in periodically could have thousands of unread notifications about build errors.

I don't feel strongly, but I am a little unsure of blog updates, which are mostly newsletters these days. The newsletters might feel like noise, especially given the blog already has a subscription for users that care about that thing.

But otherwise, the rest of this list makes sense. Maybe to help define what notifications fit in this area, these notifications should:

- Be low enough priority that we don't need to show the message in the dashboard body
- Contain content that is important enough that we want to distract the user on every dashboard page
- Never overwhelm the user. Avoid stacking more than a couple notifications per day or perhaps even avoid stacking more than 10 overall unread notifications

Reply:
I agree with most of you said here.

I was thinking about notifications I'd like to see, but also that I'd be able to enable/disable from a preference page at some point 😄 -- for example if I find that "build errors" are too noisy.

> Be low enough priority that we don't need to show the message in the dashboard body
> Contain content that is important enough that we want to distract the user on every dashboard page

I thought, for example regarding a "build tip", a workflow like the following:

- we show a notification in the build detail's page saying "we recommend you to use <something>"
- the user keeps building without following our tip
- we detect that after ~N weeks, the user keeps building without following our tip
- at this point, we create the same "build tip" notification, but we show it in the bell

> Never overwhelm the user. Avoid stacking more than a couple notifications per day or perhaps even avoid stacking more than 10 overall unread notifications

Yes. I don't think this list should grow more than a small bunch of notifications. If we want to, we can eventually create a notifications page that list all the notifications where they can be mark as read/dismissed/etc in bulk. This is out of the scope, but something to keep in mind.

----

Again, all of this needs some exploration on the UX and testing with real users to accumulate some feedback. We don't have to define how it will work. For now, we need a way to differentiate a notification that goes under the bell that another one that doesn't --which I think the `bell=` or `user=` field solves that problem.

Reply:
> also that I'd be able to enable/disable from a preference page at some point

Ah, good point! Once we have something like this, I think we could easily make build errors and opt-in notification option.

> I thought, for example regarding a "build tip", a workflow like the following

This could make sense, I was considering similar with build errors. Instead of individual build errors, we detect and notify on a state like "Hey, your project has over 10 failed consecutive builds, something is wrong."

I guess thinking about it more, we don't have to emit that error more than once either. The point of the notification system was to avoid notification deduplication. In this case, that might mean we're updating/replacing this sort of notification periodically -- like the following week "Hey, your projects has over 50 failed builds ...".

Either way, I think we're both thinking in the same direction here, and we have some options to experiment with.